### PR TITLE
perf!: reduce work to build and fix bugs in `TileLayer`, and improve culling of off-screen tiles

### PR DESCRIPTION
**Pull request by ignatz**
A couple of performance improvements:
 * cheaper data structures
 * cheaper construction of stale Tiles
 * add missing early abort to _reainChildren
 * fewer allocations
 * and just do not render tiles that are off screen.

### CODE REVIEW

**Comment by JaffaKetchup** [comment]
Just to note with the latest commit and culling tiles off screen (I thought we already did this!):
 * Be careful not to break `panBuffer` and `zoomBuffer` (I haven't tested this yet)
 * (Less importantly, one of my long term goals might change this logic again: https://github.com/fleaflet/flutter_map/projects/4#card-90964116)

File: lib/src/layer/tile_layer/tile_coordinates.dart
Comment (JaffaKetchup) [inline lib/src/layer/tile_layer/tile_coordinates.dart:18]: Not 100% a fan of this. Would prefer it if there was a separate `distanceToSq` method for more efficient internal use.

**Review by JaffaKetchup — CHANGES_REQUESTED** [review]

File: lib/src/layer/tile_layer/tile_coordinates.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_coordinates.dart:18]: No strong feelings. More importantly are you happy with me removing the distanceTo overload? :)

Comment (JaffaKetchup) [reply lib/src/layer/tile_layer/tile_coordinates.dart:18]: Not really, feels like an unnecessary breaking change.

**Comment by ignatz** [comment]
> Just to note with the latest commit and culling tiles off screen (I thought we already did this!):

Maybe you did something else? I did check the filter I added (where(visibleTileRange.contains)) and it does filter out tiles. It's not a no-op.

> Be careful not to break panBuffer and zoomBuffer (I haven't tested this yet)

It shouldn't. I'm not pruning any more aggressively.

File: lib/src/layer/tile_layer/tile_coordinates.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_coordinates.dart:18]: I understand. I'm trying to be opportunistic here. It rubs me that TileCoordinate is a Point<int> (which semantically it is not) and then overrides Point<int>'s with even more error prone behaviors. Is there a process for us to clean these up or at least stop the bleeding?

**Review by JaffaKetchup — CHANGES_REQUESTED** [review]
It appears this has broken the zooming pruning/replacement logic (this fork on the left, latest 'master' on the right):

https://github.com/fleaflet/flutter_map/assets/58115698/6a7096c3-64a0-4592-9e27-28eb6e6730df

File: lib/src/layer/tile_layer/tile_coordinates.dart
Comment (JaffaKetchup) [reply lib/src/layer/tile_layer/tile_coordinates.dart:18]: Ok, let's keep this PR non-breaking. Then in a separate PR, we can do breaking changes, with as many deprecations as possible.

Comment (JaffaKetchup) [reply lib/src/layer/tile_layer/tile_coordinates.dart:18]: @ignatz You know what, scratch that :D. We're trying to improve performance here, so breaking changes are needed. I've approved and merged your other PR with breaking changes as well.
However, in this case, I still think `distanceTo` can be useful, and so shouldn't be removed.

Apologies for the messing about, but if you're serious about not liking `Point` and it's generic types, I'll allow a PR with breaking changes.

Comment (ignatz) [reply lib/src/layer/tile_layer/tile_coordinates.dart]: No worries at all. I understand the motivations. I'm also been hit by keeping in sync with some of the v3 -> v6 major changes. I'm mostly just keeping the nag-rate high :)

**Comment by ignatz** [comment]
Great catch with the zooming. It was the render-pruning (not storage pruning). I now took the cheap way out. I checked again, there's still render-pruning happening but there's plenty of optimization left on the table.

The problem was that when zooming in the center of the parent tile wasn't within visible bounds anymore. To be correct we'd have to check for overlapping bounding boxes instead.

File: lib/src/layer/tile_layer/tile_coordinates.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_coordinates.dart]: >  However, in this case, I still think distanceTo can be useful, and so shouldn't be removed.

(I put it back. We can always defer breaking changes until v7. I do disagree though that one should support distance calculations between different types of points. Chances are it's accidental. I do think it's fair and more explicit to ask users to convert a coordinate to a Point<double> first. Either way, I'm not going to loose any sleep over this. I just think that FM will have an easier time long-term if from a separation of responsibilities point of view, it's not in the business of providing convenience on top (or extending) external types.... I'll shut up now :) )

**Comment by ignatz** [comment]
Just to roughly quanitfy the impact. I've seen roughly a 4x reduction in time spend per frame on rendering the tile-layer. For rendering only a map, overall I've seen a 30+% reduction on the ui thread.

**Comment by ignatz** [comment]
I clawed back some the performance improvements left on the table, i.e. the render pruning is more aggressive again. It's not perfect, we could be more selective (like the storage pruning) when the tiles at the target zoom-level are not yet ready.

**Comment by ignatz** [comment]
I couldn't resist. Now we're rendering the actual minimal number of tiles w/o visual regressions. No more render performance left on the table by the tile renderer.

**Comment by ignatz** [comment]
Doing a bit of qualitative testing, I must admit the savings to the render thread are not overly impressive. On the bright side, I'm happy that flutter is more efficient at drawing images than cavases :)

Panning before and after (where you'd expect render pruning to have a lesser effect):
Screenshot_20231106-142502_Wingit
Screenshot_20231106-142628_Wingit

Zooming where you'd expect a bigger effect:

Screenshot_20231106-142842_Wingit
Screenshot_20231106-143016_Wingit

:/ Makes one wonder if the extra work on the UI thread is even worth it. Happy to discuss. Will likely also depend on the specific platform

**Review by josxha — CHANGES_REQUESTED** [review]
hi @ignatz, thanks for the pull request. I haven't tested the changes yet but I like the optimisation tweaks. (:
Can you check my comments?

**Comment by ignatz** [comment]
@mootw yeah, I also didn't see a significant change in raster times either on Android either. W/o any reliable statistics, I've seen it go either way. We're definitely rendering fewer tiles though. As for the explanation why this doesn't show clear savings, I can only offer hypotheses:
 * Rendering images is super cheap
 * The more frequent changes to the render tree entail other costs that cancel out.

FWIW, the canvaskit renderer I'm actually seeing a slight improvement :shrug: 

As for my argument weighing off ui-thread vs raster-thread time, I think you misunderstood. I can leave the rastering untouched and do even less on the UI thread. I'm deliberately, did more work on the ui thread to reduce the number of tiles being rendered. However, it doesn't seem to pay off on Android in a clear way. Might be interesting to look at iOS :shrug:

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_image_view.dart:12]: These java annotations have no meaning other than expressing  constraints on sub-types. Please do not inherit from TileImageView.

I could mark it @final but if sub-typing is a concern I'd rather kill the file and make it an internal class to the tile manager.

File: lib/src/layer/tile_layer/tile_image_manager.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_image_manager.dart:19]: HashMap satifies the Map interface and is therefore iterable. I'm not sure I understand your point. If your point is that preserving the insertion order might make iteration cheaper, it doesn't

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_image_view.dart:46]: Sure, let's kill some indentation. Cleanliness is highly subjective.

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_image_view.dart:36]: Looks like a duplicate.

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_image_view.dart:52]: This one is a bit more nuanced mostly because TileImageView looks like a public API, which it really isn't. I did document this function, just on the Manager, which seemed like a more public API to me (that's debatable though. I'd rather not duplicate documents, that just creates work. If you want I can move the documentation from the manager to here.

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_image_view.dart:57]: Done

File: lib/src/layer/tile_layer/tile_layer.dart
Comment (mootw) [reply lib/src/layer/tile_layer/tile_layer.dart:754]: i think a well described simple pure function does not need a comment explaining what it does. that is just excessive. however, if we already have a distance function i think we should avoid making a new one if possible!

Comment (ignatz) [reply lib/src/layer/tile_layer/tile_layer.dart:754]: This is certainly not the most cirtical optimization but it's very common in GIS application to work with squared distances rather than distances (look at any rtree). The sqrt doesn't change the order.

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (josxha) [reply lib/src/layer/tile_layer/tile_image_view.dart]: Yes you're right, sorry. The comment didn't showed up after I reloaded the page.

File: lib/src/layer/tile_layer/tile_layer.dart
Comment (josxha) [reply lib/src/layer/tile_layer/tile_layer.dart:754]: Even through I wouldn't completely agree, I see your point. 
Therefore I'm fine with not having these private functions documented with a comment.

File: lib/src/layer/tile_layer/tile_image_manager.dart
Comment (josxha) [reply lib/src/layer/tile_layer/tile_image_manager.dart:19]: I wasn't sure if HashMap contains an Iterator when adding the comment but you're completely right. Sorry for the confusion.

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (josxha) [reply lib/src/layer/tile_layer/tile_image_view.dart:12]: While it doesn't give any performance improvements, I won't hurt performance too. 
Immutable annotations helps to mark model classes and enforce their intended use in the future through lint warnings.
In my opinion they are nice to have, especially with lint checking in the CI.

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (josxha) [reply lib/src/layer/tile_layer/tile_image_view.dart:50]: I think it's better to keep the documentation inside the manager class for the same reason that you mention.
Maybe you could make the purpose of the method more clear in the name? Like naming it `filterStaleTiles()` or even by using a getter. What do you think? (same would apply to the other filter functions of the model)

Comment (ignatz) [reply lib/src/layer/tile_layer/tile_image_view.dart:50]: Are you suggesting to change existing APIs. FWIW, I made renderTiles() match the pre-existing staleTiles(). I'm not partial to any, I'm mostly wondering who the audience is since TimeImageView is an internal API

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_image_view.dart:12]: I'm all for explicit intent that's why I'm saying @immutable is a poor choice.

The fact that you're arguing that there can be linter warnings when someone inherits from TileImageView makes me wonder if you think folks should inherit from TileImageView? If I saw an @immutable annotation, I certainly would wonder if that's what the authors wanted me to do.

I've offered you more explicit alternatives, i.e. squash the file and move TileImageView as _TileImageView into the manager's file.

**Comment by josxha** [comment]
I had some issues with displaying the correct tiles. This only happens if the tiles are loading from the web and not in memory cached.

https://github.com/fleaflet/flutter_map/assets/34318751/b4410553-735f-49a5-819f-5e5316108346

**Comment by ignatz** [comment]
Thanks for testing it. What would you expect to see? In the video it looks
like all the tiles are popping it at once rather than tile by tile. Is that
the issue you're observing?

**Comment by josxha** [comment]
> What would you expect to see?

I edited my last comment and added additional information for better understanding.

**Comment by ignatz** [comment]
Great catch and hanks for the detailed explanation, that made it crystal clear. From the video is hard to judge when the follow-up input was given.

Should be fixed now.

For context, the z-index sorting is broken at head. Funny enough it has two bugs that *almost* cancel each other out:
1. First it sorts in the wrong order. Small zIndex should go to the back and not the front for the Stack to render them on top.
2. *Fortunately*, the zIndex computation is also broken as well: maxZoom is double.infinity amking the entire sort a no-op :laughing: 

As a result, we render them in insertion order, with later tiles rendered on top due to the default hash map implemenation preserving render order. When I changed the map representation to a map that does not preserve insertion order, the broken sorting surfaced :rofl:

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (JaffaKetchup) [reply lib/src/layer/tile_layer/tile_image_view.dart:12]: I wouldn't particularly mind either way (I don't do it for my own projects), but for consistency it's probably worthwhile.

I think what @josxha's trying to say is that if you add a non-final attribute to an \@immutable marked class (for example in another PR), then the linter will immediately get you to reconsider. Less so for others inheriting in their own code. But I also don't know whether \@final does this as well.

Comment (ignatz) [reply lib/src/layer/tile_layer/tile_image_view.dart:12]: Sorry if I sound reluctant. I wanna be a good sport. I just don't believe in following the process for the sake of process. If anyone has an argument on why the immutable annotation should be there, I'll add it in a heartbeat, no questions asked.

> I think what @josxha's trying to say is that if you add a non-final attribute to an https://github.com/immutable marked class (for example in another PR), then the linter will immediately get you to reconsider. Less so for others inheriting in their own code.

The linter will do that for you for any class with a const constructor. The immutable annotation is specifically to enforce constraints on subtypes: https://api.flutter.dev/flutter/meta/immutable-constant.html.

I'm all for consistency, if this was a class that should be sub-typed and other classes/interfaces that should be sub-typed were annotated as well, I would annotate this class. Unfortunately, neither holds here and especially nobody should feel encouraged to subtype this class.

It's not even clear to me, what a good use-case for @Immutable would look like. Let's say, we defined a base-class for folks to extend. Why would we disallow our users, the implementers, to have a constructor or have mutable members for caching? Maybe I could imagine a case, where we'd manage a complex lifecycle of the object and we wouldn't want implementation to implicitly depend on it :shrug: 

I added a final class modifier now, which will at least disallow sub-typing from outside this package. I still think the most consistent would be to make it a private class, since users should neither sub-type it nor look at it.

(and completely tangentially, I believe that as we push performance further and further, for any GC optimizatoins constness/immutably is here to bite us. Right now we're constructing thousands of TileCoordinates per frame because TileCoordinate is const (immutable or not))

Comment (josxha) [reply lib/src/layer/tile_layer/tile_image_view.dart:12]: > I think what @josxha's trying to say is that if you add a non-final attribute to an https://github.com/immutable marked class (for example in another PR), then the linter will immediately get you to reconsider.

Yes, indeed, my intention here was to ensure that if someone wants to add mutable fields in the future in pull requests they will get blocked by a warning. I haven't had polymorphism in mind at all. Now I understand your reason you don't want to add it much better. Sorry for the confusion and thanks for bearing with me.

Closing this this as resolved.

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (josxha) [reply lib/src/layer/tile_layer/tile_image_view.dart:50]: > Are you suggesting to change existing APIs?

Yes. To make it better understandable to someone who dives into the code (code audit, future contributor, ...) that for example `renderTiles()` doesn't mean to render tiles but to filter and return the tiles that should get rendered.

However, if you don't want changes to existing APIs as part of your pull request, I can understand that too. Then leave it as is.

**Comment by josxha** [comment]
> Funny enough it has two bugs that almost cancel each other out.

Thats somehow kind of impressive. 
I tested again and couldn't find any more issues. You can decide how to resolve the last change request. After that it's ready to merge.

File: lib/src/layer/tile_layer/tile_image_view.dart
Comment (ignatz) [reply lib/src/layer/tile_layer/tile_image_view.dart]: > hat for example renderTiles() doesn't mean to render tiles but to filter and return the tiles that should get rendered.

Good point. Render is the wrong verb for  what the method does. Changed

**Comment by ignatz** [comment]
I snuck in one more minor change. I merged `TileImageManager.(createMissingTiles|createMissingTilesIn)` to have fewer code paths doing the same thing but I'm also hoping the reduce the number of TileCoordinates created in a follow-up that we're not spamming the GC with thousands of coordinates per frame.

Thanks for the review. Great catch on the tile stacking order.

**Review by josxha — APPROVED** [review]
lgtm!

**Comment by JaffaKetchup** [comment]
Ok, I'll review this tomorrow as well.
Would also appreciate it if we could whittle down this PR/commit title, but not sure how we're going to do that (also, it looks like we've begun to follow conventional commits, so it must be fix or perf, not both, but this isn't too important).

**Comment by ignatz** [comment]
Thanks for the review.

> also, it looks like we've begun to follow conventional commits, so it must be fix or perf, not both

:rofl: great example of process getting in the way :hide:

**Pull request merged**
